llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -2307,9 +2309,11 @@ Value *InstCombiner::getFreelyInvertedImpl(Value *V, bool WillInvertAllUses,
   // If `V` is of the form `A ^ ~B` then `~(A ^ ~B)` can be folded
   // into `A ^ B` if we are willing to invert all of the uses.
   if (match(V, m_Xor(m_Value(A), m_Value(B)))) {
+    bool DoesConsumeOldValue = DoesConsume;
     if (auto *BV = getFreelyInvertedImpl(B, B->hasOneUse(), Builder,
                                          DoesConsume, Depth))
       return Builder ? Builder->CreateXor(A, BV) : NonNull;
+    DoesConsume = DoesConsumeOldValue;
     if (auto *AV = getFreelyInvertedImpl(A, A->hasOneUse(), Builder,
                                          DoesConsume, Depth))
       return Builder ? Builder->CreateXor(AV, B) : NonNull;
```

Comment:
Still trying to properly understand why this is the fix, but something to note:
below at line:2377 seems buggy.

```
   if (NewIncomingVal == V)
        return nullptr;
```

`NewIncomingVal` uses `builder` as `nullptr` so if the return is non-null, it will return `NonNull` which will never equal `V`

Reply:
See also the comment https://github.com/llvm/llvm-project/pull/80804#discussion_r1481123325.

Reply:
think the `Builder == nullptr` case should return `V` (which is equally valid as `NonNull`).
I see that it ends up working b.c the force of recursive depth to the end and `/*WillInvertAllUses=*/false` will make it only match `not`. But should be fixed up I think.

Not related to the current bug though.

Reply:
At the very least should add an assert that `NewIncomingVal != NonNull`

Reply:
> Still trying to properly understand why this is the fix, but something to note: below at line:2377 seems buggy.
> 
> ```
>    if (NewIncomingVal == V)
>         return nullptr;
> ```
> 
> `NewIncomingVal` uses `builder` as `nullptr` so if the return is non-null, it will return `NonNull` which will never equal `V`

I believe it never happens since `NewIncomingVal` always evaluates to nullptr if `U.get()` is also a phi node.

Reply:
I don't think the code is explicitly buggy now, but it seems like an accident waiting to happen. Although it won't be silent (a v explicit segfault), so guess there is no real alarm.